### ISSUE TITLE
[discussion] Move semantics for tensors

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#106596]
### 🚀 The feature, motivation and pitch

Sometimes we want to kind of pass the tensor to a function entirely (so that it can safely do inplace modifications without worrying to disturb other consumers).

One way it could work is that after a new_tensor_ref = tensor.move() is called, the previous tensor reference becomes somewhat invalid and would throw exception on any usage.

There could also be some `assert tensor.is_single_owner()` which could ensure that there is only one reference to the tensor.

Of course, there would be ways to circumvent this (e.g. by recording manually data_ptr), but it could be a way to increase expressiveness, give compiler more hints for compilation, and let the user express  clearer ownership semantics and have some guards.

**Comment by vadimkantorov** [comment]
The related issue is https://github.com/pytorch/pytorch/issues/46948 on using assert statements to give shape prop hints to the compiler. I propose to use more explicit compiler hints (for torch.compile) and assert statements as a way to encode these hints naturally

**Comment by albanD** [comment]
Actually, I was wondering if it would be possible to do a move() in python such that:
```python
def foo(inp):
  interm = torch.mul(move(inp), 2)
  # The Tensor for inp was de-allocated on the line above just after the call to mul().
  interm2 = torch.add(interm, 3)
  return interm2
out = foo(move(inp))
# inp is undefined now
```

That could help with reducing memory in general where you could move() your network input or other intemediary into the function and it will be freed "eagerly" within that function.

But not sure if that is feasible in python...

**Comment by vadimkantorov** [comment]
I think we can only emulate it. E.g. make it so that any other method on `inp` throws after it was "moved" - so we can't support this semantics in the compiler, only by runtime and consider it as a strong hint to the compiler / runtime

**Comment by vadimkantorov** [comment]
I guess the move-semantics can also be helpful for this can-return-view/force-copy aspect we discussed here: https://github.com/pytorch/pytorch/issues/106584#issuecomment-1666483832

**Comment by albanD** [comment]
btw a very good comment from Zach: You can mimick this by using list and pop:
```python
inp_holder = [inp]
out = foo(inp_holder.pop())
```

**Comment by vadimkantorov** [comment]
Yeah, one can implement some sort of holder like this, but for to be actually user, I would suppose that the widest used type of torch.Tensor should support this directly...

**Comment by albanD** [comment]
The problem is that our `torch.Tensor` type is not really a holder... It IS the Tensor. So I'm not sure how feasible that is...

**Comment by jon-chuang** [comment]
Hello @vadimkantorov , seems we have a similar idea about move semantics for buffers to enable more aggressive in-place optimizations.

I reproduce my idea here:

---

As a more long-term observation, if inductor/AOTAutograd is allowed to take ownership of inputs, it could be more aggressive in the bufferization optimizations. This could involve some kind of `move` keyword in pytorch.

Example API:
```python
@torch.compile(move_inputs=True)
class MyModel(nn.Module):
  ..

x = torch.tensor(..)
m = MyModel()
y = m(x)

x.__getattr__(..) # raise `AccessingMovedValue`
```

Even better if is Python exposed some kind of move semantics so that implictly:
```python
y = m(x); x = None
```

**Comment by jon-chuang** [comment]
Overall, I think this optimization's pratical effect would be limited:

1. A single copy of the input suffices to ensure that it is owned by the graph for a deep network. Hence, this copied buffer can be reused many times
2. For optimizer updates, one already assumes ownership and performs in-place updates.

Hence, I believe that a more productive direction, is in **module-level bufferization** _across torch dynamo fx graphs in a single module_. Some observations are detailed in the `details` in this comment: https://github.com/pytorch/pytorch/issues/109240#issuecomment-1737764251

**Comment by vadimkantorov** [comment]
For (1), in general you are right, but maybe ways of ownerships checking / check if there're any hooks and so on, can be used for optimizing the eager mode as well. And in training / eager, sometimes having one more copy is hurting memory, so being able to express user hints would be nice! (also for improving code clarity)
